src/varnam_state.cpp

```diff
@@ -368,49 +368,79 @@ void VarnamState::selectCandidate(int index) {
+    bool enablePunctuation = engine_->getConfig()->enablePunctuation.value();
+
+    if (isWordBreakKey) {
+        if (candidates && candidates->cursorIndex() >= 0) {
+            stringToCommit = candidates->candidate(candidates->cursorIndex()).text().toStringForCommit();
+        } else {
+            stringToCommit = preedit_.toStringForCommit();
+        }
+
+        if (enablePunctuation) { // If punctuation processing is enabled
+            buffer_.clear(); // Clear buffer
+            buffer_.push_back(getWordBreakChar(key)[0]); // Add word break char to buffer
+            
+            // Concatenate result word if available
+            if (getVarnamResult() && result_ && varray_length(result_) > 0) {
+                vword *result_word = static_cast<vword *>(varray_get(result_, 0));
+                if (result_word) {
+                    stringToCommit = stringutils::concat(stringToCommit, result_word->text);
+                }
+            }
+        } else {
+            // If punctuation processing is not enabled
+            // Append the punctuation character to stringToCommit
+            stringToCommit += getWordBreakChar(key);
+            // Set wordToLearn to the base candidate text without punctuation
+            wordToLearn = candidates->candidate(candidateSelected).text().toStringForCommit();
+        }
+    } else if (key == FcitxKey_Escape || key == FcitxKey_0 ||
+               !candidates || candidates->size() <= 1 || !result_ || varray_is_empty(result_)) {
+        // Handle escape or special keys
+        stringToCommit.assign(preedit_.toStringForCommit());
+        candidateSelected = 0;
+    } else if ((candidates->cursorIndex() <= 0) && !candidateSelected) {
+        // Handle the case where no candidate is selected
+        vword *first_result = static_cast<vword *>(varray_get(result_, 0));
+        if (first_result) {
+            stringToCommit.assign(first_result->text);
+            candidateSelected = 1;
+        }
+    } else {
+        // Handle regular candidate selection
+        stringToCommit.assign(candidates->candidate(candidateSelected).text().toStringForCommit());
     }
-  } else {
-    stringToCommit.assign(
-        candidates->candidate(candidateSelected).text().toStringForCommit());
-  }
 
-  if (isWordBreak(key)) {
-    stringToCommit = stringutils::concat(stringToCommit, getWordBreakChar(key));
-  }
+      wordToLearn = stringToCommit;
 #ifdef DEBUG_MODE
-  VARNAM_INFO() << "string to commit:" << stringToCommit << " "
-                << getWordBreakChar(key);
+    VARNAM_INFO() << "string to commit:" << stringToCommit << " "
+                  << getWordBreakChar(key);
 #endif
-  ic_->commitString(stringToCommit);
+    ic_->commitString(stringToCommit);
+if (stringToCommit.empty() || lastTypedCharIsDigit || isWordBreakKey ||
+        ic_->capabilityFlags().test(CapabilityFlag::PasswordOrSensitive) ||
+        !engine_->getConfig()->shouldLearnWords.value() || !candidateSelected) {
```

Comment:
Check if the learn word function is getting called if the last entered character is a punctuation mark.

Reply:
i think line https://github.com/khumnath/varnam-fcitx5/blob/bce13c04f84e52332b20efd4dd2def5b362f7af0/src/varnam_state.cpp#L438 should call learning function when punctuation mark is typed.

Reply:
Check whether the mentioned line is getting executed or not

Reply:
When I type a word like ‘प्रवन्ध’ following with word break key or a digit, the word is learned. I haven’t tested it in other situations, but it should work as expected based on this code.